docs/writing_plugins.md

```diff
@@ -0,0 +1,550 @@
+
+```ts
+parseOption(key: string, value: string | undefined): void;
+```
+
+This function will be invoked by the framework, passing in any key/value pairs that it does not recognize from its pre-defined list.
+
+### Reading custom options
+
+Protobuf-ES does not yet provide support for extensions, neither in general as pertaining to proto2 nor with custom options in proto3.  However, in the interim, there are convenience functions for retrieving any custom options specified in your .proto files.  These are provided as a temporary utility until full extension support is implemented.  There are three functions depending on the structure of the custom option desired (scalar, message, or enum):
+
+#### Scalar Options
+
+Custom options of a scalar type (`boolean`, `string`, `int32`, etc.) can be retrieved via the `findCustomScalarOption` function.  It returns a type corresponding to the given `scalarType` parameter.  For example, if `ScalarType.STRING` is passed, the return type will be a `string`.  If the option is not found, it returns `undefined`.
+
+```ts
+function findCustomScalarOption<T extends ScalarType>(
+  desc: AnyDesc,
+  id: number,
+  scalarType: T
+): ScalarValue<T> | undefined;
+```
+
+`AnyDesc` represents any of the `DescXXX` objects such as `DescFile`, `DescEnum`, `DescMessage`, etc.  The `id` parameter represents the ID of the custom options field definition.
+
+The `scalarType` parameter is the type of the custom option you are searching for.  `ScalarType` is an enum that represents all possible scalar types in the Protobuf grammar
+
+For example, given the following:
+
+```proto
+extend google.protobuf.MessageOptions {
+  optional int32 foo_message_option = 50001;
+}
+extend google.protobuf.FieldOptions {
+  optional string foo_field_option = 50002;
+}
+
+message FooMessage {
+  option (foo_message_option) = 1234;
+
+  int32 foo = 1 [(foo_field_option) = "test"];
+}
+```
+
+The values of these options can be retrieved as follows:
+
+```ts
+const msgVal = findCustomScalarOption(descMessage, 50001, ScalarType.INT32);  // 1234
+
+const fieldVal = findCustomScalarOption(descField, 50002, ScalarType.STRING);  // "test"
+```
+
+#### Message Options
+
+Custom options of a more complex message type can be retrieved via the `findCustomMessageOption` function.  It returns a concrete type with fields populated corresponding to the values set in the proto file.
+
+```ts
+export function findCustomMessageOption<T extends Message<T>>(
+  desc: AnyDesc,
+  id: number,
+  msgType: MessageType<T>
+): T | undefined {
+```
+
+The `msgType` parameter represents the type of the message you are searching for.  Note that this can be a message generated from a proto file or can be a type created at runtime via `makeMessageType`.
```

Comment:
Do we need to call out that you can also pass a type created at run time? 

Our recommendation should be to generate code for this message, because it is much more convenient.